Do not replace existing IPC socket

psychon opened the pull request:
Imagine you are a happy i3 user and want to also write patches for i3.
You use "Xephyr :1" to get another X11 server and then start your newly
build i3 in it with "DISPLAY=:1 ./i3". You test your changes and
everything seems fine. You are happy. Later that day, you try to log
out, but the $mod+Shift+e key binding from the default config no longer
works. i3-msg cannot connect to the IPC socket because "No such file or
directory". What is going on?

The problem boils down to $I3SOCK having something like two meanings.
When i3 starts, it sets the environment variable $I3SOCK to the path of
its IPC socket. That way, any process started from i3 inherits this and
i3-msg knows how to talk to i3. However, when this variable is already
set when i3 starts, then i3 will replace the existing socket. Thus, in
the earlier experiments, the "separate i3" that was used for
experimenting stole the "main i3"'s socket and replaced it with its own.
When it exited, it deleted that socket.

This commit adds half a work around to this problem: When creating the
IPC socket, i3 will now first try to connect() to the socket. If this
succeeds, it will complain and refuse to use this socket. If the
connect() call fails, it will proceed as usual and create the socket.

Note that trying to connect() to a socket that no process listens on
will fail. Thus, this new code only really "triggers" when some process
is actively listening on this socket and accepting connections.

Example output for when the socket is already in use:

$ I3SOCK=/tmp/sdfdsf DISPLAY=:2 ./i3
31.10.2021 17:03:55 - [libi3] ERROR: Refusing to create UNIX socket at /tmp/sdfdsf: Socket is already in use
31.10.2021 17:03:55 - ERROR: Could not create the IPC socket, IPC disabled

This commit sadly only provides part of the solution. i3 will still
delete the socket when shutting down, even if it failed to create the
IPC socket. Thus, the ipc socket will still break, but now only later.
This will be fixed separately.

First-step-towards-fixing: https://github.com/i3/i3/issues/4381
Signed-off-by: Uli Schlachter <psychon@znc.in>

----

Only unlink() IPC socket if we created it

When the IPC socket could not be created, i3 would still unlink() the
socket during shutdown. This might mean that "unrelated stuff" is
removed. In the issue below, the IPC socket of another i3 instance was
unlinked.

Together with the previous commit, this fixes
https://github.com/i3/i3/issues/4381

Signed-off-by: Uli Schlachter <psychon@znc.in>

psychon committed:
Do not replace existing IPC socket

Imagine you are a happy i3 user and want to also write patches for i3.
You use "Xephyr :1" to get another X11 server and then start your newly
build i3 in it with "DISPLAY=:1 ./i3". You test your changes and
everything seems fine. You are happy. Later that day, you try to log
out, but the $mod+Shift+e key binding from the default config no longer
works. i3-msg cannot connect to the IPC socket because "No such file or
directory". What is going on?

The problem boils down to $I3SOCK having something like two meanings.
When i3 starts, it sets the environment variable $I3SOCK to the path of
its IPC socket. That way, any process started from i3 inherits this and
i3-msg knows how to talk to i3. However, when this variable is already
set when i3 starts, then i3 will replace the existing socket. Thus, in
the earlier experiments, the "separate i3" that was used for
experimenting stole the "main i3"'s socket and replaced it with its own.
When it exited, it deleted that socket.

This commit adds half a work around to this problem: When creating the
IPC socket, i3 will now first try to connect() to the socket. If this
succeeds, it will complain and refuse to use this socket. If the
connect() call fails, it will proceed as usual and create the socket.

Note that trying to connect() to a socket that no process listens on
will fail. Thus, this new code only really "triggers" when some process
is actively listening on this socket and accepting connections.

Example output for when the socket is already in use:

$ I3SOCK=/tmp/sdfdsf DISPLAY=:2 ./i3
31.10.2021 17:03:55 - [libi3] ERROR: Refusing to create UNIX socket at /tmp/sdfdsf: Socket is already in use
31.10.2021 17:03:55 - ERROR: Could not create the IPC socket, IPC disabled

This commit sadly only provides part of the solution. i3 will still
delete the socket when shutting down, even if it failed to create the
IPC socket. Thus, the ipc socket will still break, but now only later.
This will be fixed separately.

First-step-towards-fixing: https://github.com/i3/i3/issues/4381
Signed-off-by: Uli Schlachter <psychon@znc.in>

orestisfl commented:
Do we realistically ever want to run i3 without IPC? @Airblader @stapelberg  it's currently allowed in 

https://github.com/i3/i3/blob/e4d3ec92172380f0ee278541b6e7356307e124ed/src/main.c#L992-L994

but do we ever want to have this happen in this day and age?

Airblader commented:
I can't imagine how such an instance of i3 would be useful at all.

psychon commented:
Rebased on `next` for #4647. Dropped commit e225b2c27a306c911583995b334a5900732d9471 "Only unlink() IPC socket if we created it"  since it is no longer relevant.

I did not re-test this change since I expect nothing changed.

orestisfl commented on libi3/ipc_connect.c:
```diff
@@ -39,6 +40,19 @@ int ipc_connect(const char *socket_path) {
     }
 
+    int sockfd = ipc_connect_impl(path);
+    if (sockfd < 0)
+        err(EXIT_FAILURE, "Could not connect to i3 on socket %s", path);
```
```suggestion
    if (sockfd < 0) {
        err(EXIT_FAILURE, "Could not connect to i3 on socket %s", path); }
```

orestisfl reviewed (changes requested):
LGTM, one nitpick

orestisfl commented:
Ok, please squash everything and I'll merge

stapelberg commented:
Thanks!

stapelberg commented:
This doesn’t build?
```
 ../libi3/create_socket.c: In function ‘create_socket’:
  ../libi3/create_socket.c:43:9: error: ‘errno’ undeclared (first use in this function); did you mean ‘errx’?
           errno = EEXIST;
           ^~~~~
           errx
  ../libi3/create_socket.c:43:9: note: each undeclared identifier is reported only once for each function it appears in
  ../libi3/create_socket.c:43:17: error: ‘EEXIST’ undeclared (first use in this function); did you mean ‘ATEXIT’?
           errno = EEXIST;
                   ^~~~~~
                   ATEXIT
```

Not sure why this only fails in CI for the full build, not for the PR build.

psychon commented:
Sorry for the breakage.

Cannot reproduce: Current `next` works for me also after the merge.

Random guess: Something else already includes `errno.h` for me but not for you? Could you try adding `#include <errno.h>`, checking if that fixes things and then just commit that to `next` yourself.

stapelberg commented:
For the record, the breakage is fixed with PR #4649
